File: backend/shared/common-security/src/main/java/tn/bewerbi/common/security/PasswordStrength.java
package tn.bewerbi.common.security;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Java port of {@code shared/lib/password-strength.ts}. Same scoring rubric
 * and suggestion IDs — see {@code shared/lib/README.md} for parity rules.
 *
 * <p>Used by identity-service to reject obviously weak passwords on
 * {@code /register} and {@code /password/reset} with a 422 carrying a
 * structured suggestion list, so the client can localise the hints.
 */
public final class PasswordStrength {

    /** Stable result. Score is in [0,4]; label is the matching stable id. */
    public record Result(int score, String label, List<String> suggestions) {}

    private static final Set<String> COMMON = Set.of(
            "password", "passwort", "123456", "12345678",
            "qwerty", "azerty", "abc123", "letmein", "iloveyou",
            "admin", "welcome", "monkey", "dragon",
            "bewerbi", "tunisia", "deutschland");

    private static final Pattern LOWER = Pattern.compile("[a-z]");
    private static final Pattern UPPER = Pattern.compile("[A-Z]");
    private static final Pattern DIGIT = Pattern.compile("[0-9]");
    private static final Pattern SYMBOL = Pattern.compile("[^A-Za-z0-9]");
    private static final Pattern REPEAT = Pattern.compile("(.)\\1\\1");

    private PasswordStrength() {}

    public static Result evaluate(String input) {
        String value = input == null ? "" : input;
        List<String> suggestions = new ArrayList<>(5);

        int raw = 0;
        if (value.length() >= 8) raw += 1;
        if (value.length() >= 12) raw += 1;

        int classes = classesPresent(value);
        if (classes >= 3) raw += 1;
        if (classes == 4 && value.length() >= 10) raw += 1;

        boolean seq = hasSequentialRun(value);
        boolean rep = hasRepeatRun(value);
        if (seq || rep) raw -= 1;

        boolean common = COMMON.contains(value.toLowerCase());
        if (common) raw -= 1;

        if (value.length() < 8) suggestions.add("length");
        if (classes < 3) suggestions.add("mixClasses");
        if (seq) suggestions.add("noSequential");
        if (rep) suggestions.add("noRepeats");
        if (common) suggestions.add("notCommon");

        int score = Math.max(0, Math.min(4, raw));
        return new Result(score, labelFor(score), List.copyOf(suggestions));
    }

    private static int classesPresent(String s) {
        int c = 0;
        if (LOWER.matcher(s).find()) c++;
        if (UPPER.matcher(s).find()) c++;
        if (DIGIT.matcher(s).find()) c++;
        if (SYMBOL.matcher(s).find()) c++;
        return c;
    }

    private static boolean hasSequentialRun(String s) {
        for (int i = 0; i < s.length() - 2; i++) {
            int a = s.charAt(i);
            int b = s.charAt(i + 1);
            int c = s.charAt(i + 2);
            if (b - a == 1 && c - b == 1) return true;
        }
        return false;
    }

    private static boolean hasRepeatRun(String s) {
        return REPEAT.matcher(s).find();
    }

    private static String labelFor(int score) {
        return switch (score) {
            case 0 -> "very-weak";
            case 1 -> "weak";
            case 2 -> "fair";
            case 3 -> "strong";
            default -> "very-strong";
        };
    }
}

File: flutter/lib/services/password_strength.dart
// @generated_parity: shared/lib/password_strength.dart
//
// Cross-platform password-strength heuristic. This file mirrors the
// canonical Dart implementation under shared/lib/ — kept in-tree so the
// Flutter build doesn't need a custom path: package or build-time copy.
// When you change the canonical file, copy the new version verbatim.
// (Long-term: a tiny shared Dart package will replace this duplicate.)

class PasswordStrengthResult {
  const PasswordStrengthResult({
    required this.score,
    required this.label,
    required this.suggestions,
  });

  final int score;
  final String label;
  final List<String> suggestions;
}

const _common = <String>{
  'password',
  'passwort',
  '123456',
  '12345678',
  'qwerty',
  'azerty',
  'abc123',
  'letmein',
  'iloveyou',
  'admin',
  'welcome',
  'monkey',
  'dragon',
  'bewerbi',
  'tunisia',
  'deutschland',
};

int _classesPresent(String input) {
  var c = 0;
  if (RegExp(r'[a-z]').hasMatch(input)) c++;
  if (RegExp(r'[A-Z]').hasMatch(input)) c++;
  if (RegExp(r'[0-9]').hasMatch(input)) c++;
  if (RegExp(r'[^A-Za-z0-9]').hasMatch(input)) c++;
  return c;
}

bool _hasSequentialRun(String input) {
  for (var i = 0; i < input.length - 2; i++) {
    final a = input.codeUnitAt(i);
    final b = input.codeUnitAt(i + 1);
    final c = input.codeUnitAt(i + 2);
    if (b - a == 1 && c - b == 1) return true;
  }
  return false;
}

bool _hasRepeatRun(String input) => RegExp(r'(.)\1\1').hasMatch(input);

String _labelFor(int score) {
  switch (score) {
    case 0:
      return 'very-weak';
    case 1:
      return 'weak';
    case 2:
      return 'fair';
    case 3:
      return 'strong';
    default:
      return 'very-strong';
  }
}

PasswordStrengthResult evaluatePassword(String input) {
  final value = input;
  final suggestions = <String>[];

  var raw = 0;
  if (value.length >= 8) raw += 1;
  if (value.length >= 12) raw += 1;

  final classes = _classesPresent(value);
  if (classes >= 3) raw += 1;
  if (classes == 4 && value.length >= 10) raw += 1;

  if (_hasSequentialRun(value) || _hasRepeatRun(value)) raw -= 1;
  if (_common.contains(value.toLowerCase())) raw -= 1;

  if (value.length < 8) suggestions.add('length');
  if (classes < 3) suggestions.add('mixClasses');
  if (_hasSequentialRun(value)) suggestions.add('noSequential');
  if (_hasRepeatRun(value)) suggestions.add('noRepeats');
  if (_common.contains(value.toLowerCase())) suggestions.add('notCommon');

  final score = raw.clamp(0, 4);
  return PasswordStrengthResult(
    score: score,
    label: _labelFor(score),
    suggestions: suggestions,
  );
}
